src/utils/bundle_packer.ts

```diff
@@ -0,0 +1,107 @@
+import { ByteCount, UploadStats } from '../types';
+
+export type BundleIndex = number;
+
+interface DataItemPlan {
+	// Upload stats can be undefined when a metaData for an over-sized file is packed
+	uploadStats?: UploadStats;
+	byteCountAsDataItems: ByteCount;
+	numberOfDataItems: number;
+}
+
+export abstract class BundlePacker {
+	constructor(protected readonly maxBundleSize: ByteCount, protected readonly maxDataItemLimit: number) {
+		if (!Number.isFinite(maxDataItemLimit) || !Number.isInteger(maxDataItemLimit) || maxDataItemLimit < 2) {
+			throw new Error('Maximum data item limit must be an integer value of 2 or more!');
+		}
+	}
+
+	public abstract bundles: PlannedBundle[];
```

Comment:
```suggestion
	public abstract readonly bundles: PlannedBundle[];
```
`readonly` will give us most of what we were actually looking for when I suggested leaving memory allocation to the subtypes and providing a getter. And in fact, it works fine with LowestIndexBundlePacker having the protected member and the public getter as you've already done. There should't be any reason for a client of this class to change the reference to the array externally.

Reply:
3dba5ea